### PR TITLE
Plfm 5086

### PR DESCRIPTION
**Pull request by john-hill**

### CODE REVIEW

File: lib/jdomodels/src/main/resources/schema/DownloadListItem-ddl.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE IF NOT EXISTS `DOWNLOAD_LIST_ITEM` (
+  `PRINCIPAL_ID` bigint(20) NOT NULL,
+  `ASSOCIATED_OBJECT_ID` bigint(20) NOT NULL,
+  `ASSOCIATED_OBJECT_TYPE` enum ('FileEntity') NOT NULL,
+  `FILE_HANDLE_ID` bigint(20) NOT NULL,
+  PRIMARY KEY (`PRINCIPAL_ID`, `ASSOCIATED_OBJECT_ID`, `ASSOCIATED_OBJECT_TYPE`, `FILE_HANDLE_ID`),
```
Comment (john-hill) [inline lib/jdomodels/src/main/resources/schema/DownloadListItem-ddl.sql:6]: add index on principal ID.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/file/download/BulkDownloadDAOImplTest.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+		list.setEtag("etag");
+		list.setPrincipalId(userId);
+		list.setUpdatedOn(System.currentTimeMillis());
+		DBODownloadListItem item = new DBODownloadListItem();
+		item.setAssociatedObjectId(123L);
+		item.setAssociatedObjectType(FileHandleAssociateType.FileEntity.name());
+		item.setFileHandleId(456L);
+		List<DBODownloadListItem> items = Lists.newArrayList(item);
+		// call under test
+		DownloadList result = BulkDownloadDAOImpl.createFromDBO(list, items);
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/file/download/BulkDownloadDAOImplTest.java:152]: rename to translateFromDBOtoDTO

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/file/download/BulkDownloadDAOImplTest.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+		DownloadList list = bulkDownlaodDao.getUsersDownloadList(userIdString);
+		assertNotNull(list);
+		assertNotNull(list.getUpdatedOn());
+		assertNotNull(list.getEtag());
+		assertEquals(null, list.getFilesToDownload());
+		assertEquals(this.userIdString, list.getOwnerId());
+	}
+	
+	@Test (expected=IllegalArgumentException.class)
+	public void testGetUsersDownloadListNullUserer() {
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/file/download/BulkDownloadDAOImplTest.java:194]: User

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/file/download/BulkDownloadDAOImplTest.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+		Thread.sleep(10L);
+		assertNotNull(old);
+		// call under test
+		bulkDownlaodDao.touchUsersDownloadList(userId);
+		
+		DownloadList updated = bulkDownlaodDao.getUsersDownloadList(userIdString);
+		assertNotNull(updated.getEtag());
+		assertNotNull(updated.getUpdatedOn());
+		assertFalse(updated.getEtag().equals(old.getEtag()));
+		assertTrue(updated.getUpdatedOn().getTime() > old.getUpdatedOn().getTime());
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/file/download/BulkDownloadDAOImplTest.java:297]: add test to touch after row exists

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/file/download/BulkDownloadDAOImplTest.java [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+		// Add a few files to a user's list
+		List<FileHandleAssociation> toAdd = fileHandleAssociations.subList(1, 3);
+		// call under test
+		DownloadList result = bulkDownlaodDao.addFilesToDownloadList(userIdString, toAdd);
+		assertNotNull(result);
+		assertEquals(userIdString, result.getOwnerId());
+		assertNotNull(result.getEtag());
+		assertNotNull(result.getUpdatedOn());
+		assertEquals(toAdd, result.getFilesToDownload());
+	}
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/file/download/BulkDownloadDAOImplTest.java:311]: add test add the same duplicate file to the same user.
add test same file to multiple users.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/migration/MigratableTableDAOImplAutowireTest.java [hunk 6]
```diff
@@ -220,6 +220,7 @@ public void testGetPrimaryMigrationTypes(){
 		expectedPrimaryTypes.add(MigrationType.DATA_ACCESS_REQUEST);
 		expectedPrimaryTypes.add(MigrationType.DATA_ACCESS_SUBMISSION);
 		expectedPrimaryTypes.add(MigrationType.DATA_ACCESS_SUBMISSION_SUBMITTER);
+		expectedPrimaryTypes.add(MigrationType.DOWNLOAD_LIST);
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/migration/MigratableTableDAOImplAutowireTest.java:223]: replace with a test that fails if you forget to register a new type. (remove test that files if you do the right thing).

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/migration/MigrationType.json [hunk 7]
```diff
@@ -301,6 +301,12 @@
         {
             "name": "DATA_ACCESS_SUBMISSION_STATUS",
             "description": "TODO: Auto-generated description"
+        },        {
+            "name": "DOWNLOAD_LIST",
+            "description": "The root object for a user's download list."
+        },        {
```
Comment (john-hill) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/migration/MigrationType.json:307]: format
